File: extract_html_text.py

# Parse useful text from USAgov static site html files into files containing
# the relevant data from each html file, without the html stuff - just a line
# of text for each relevant html element


from bs4 import BeautifulSoup
from pathlib import Path
import re
import os

skip_text = [
    "Find an office near you:",
    "Contact:",
    "Website:",
    "Phone number:",
    "Ask a real person any government-related question for free. They will get you the answer or let you know where to find it.",
    "SHARE THIS PAGE:",
    "LAST UPDATED:" ]

# Name the css classes which are most likely to contain useful text
css_classes = ['usa-prose', 'usa-card__body', 'life-events-item-content',
               'usagov-directory-table']

root_path = '.'
in_path = root_path + '/input'
out_path = root_path + '/output'

# recursively get the path of each html file we want to process for input
html_files = []
for root, _, files in os.walk(in_path):
    for file in files:
        if file.endswith(".html"):
            html_files.append(os.path.join(root, file))

# iterate over each of the input html files, extracting and writing what we
# need into a corresponding output file
for html_file in html_files:
    with open(html_file, 'r', encoding='utf-8') as file:
        html_cont = file.read()

    # each static file in USAgov is named index.html, and lives
    # in an appropriately named directory.  We want to take that
    # directory name, and use it as the filename for the output
    # file, rather than having a directory per output file
    fin_path, output_file = os.path.split(html_file)
    path_parts = Path(fin_path).parts
    page_path = path_parts[-1]
    output_file, _ = os.path.splitext(output_file)
    output_file = out_path + '/' + page_path + '.dat'

    # ping user
    print("Processing to '" + output_file + "'")

    items_written = 0
    # open the receiving file, before we start processing the input file
    check_duplicates = []
    with open(output_file, 'w', encoding='utf-8') as ofile:
        soup = BeautifulSoup(html_cont, 'html.parser')

        divs = soup.find_all('div', class_=css_classes)
        for div in divs:
            # find all the sub-elements we're interested in
            paragraphs = div.find_all('p')
            spans = div.find_all('span')
            hrefs = div.find_all('href')
            items_of_interest = paragraphs + spans + hrefs

            # iterate over each sub-element
            for i in items_of_interest:
                s = i.get_text()
                if s not in check_duplicates:
                    check_duplicates.append(s)
                    # remove boilerplate text
                    process = True
                    for skip in skip_text:
                        if s.startswith(skip):
                            process = False
                    if process:
                        # cleanup whitespace, etc
                        s = re.sub(r"\s+|\r+|\n+|\t+", " ", s)
                        s.join(s.split())
                        # write to output file
                        print(s, file=ofile)
                        items_written += 1

    # close output file before proceeding to the next input file
    ofile.close()
    if items_written == 0:
        print("No data written for '" + output_file + "'. Removing file")
        os.remove(output_file)

File: php/extract-html-text.php
<?php
// Parse useful text from USAgov static site html files into files containing
// the relevant data from each html file, without the html stuff - just a line
// of text for each relevant html element

$SKIP_TEXT = [
    "Find an office near you:",
    "Contact:",
    "Website:",
    "Phone number:",
    "Ask a real person any government-related question for free. They will get you the answer or let you know where to find it.",
    "SHARE THIS PAGE:",
    "LAST UPDATED:"
];

$CSS_CLASSES = ['usa-prose', 'usa-card__body', 'life-events-item-content', 'usagov-directory-table'];

$ROOT_PATH = '..';
$INPUT_PATH = $ROOT_PATH . '/input';
$OUTPUT_PATH = $ROOT_PATH . '/output';

$EXCLUDE_DIRS = ['es', 'espanol', 'sites', 'core', 'modules', 'themes', 's3', '_data'];

function getHtmlFiles($dir, $excludeDirs) {
    $rii = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dir));
    $files = [];
    foreach ($rii as $file) {
        if ($file->isDir()) continue;
        $filePath = $file->getPathname();
        $parts = explode(DIRECTORY_SEPARATOR, $file->getPath());
        if (array_intersect($parts, $excludeDirs)) continue;
        if (substr($filePath, -5) === ".html") {
            $files[] = $filePath;
        }
    }
    return $files;
}

$html_files = getHtmlFiles($INPUT_PATH, $EXCLUDE_DIRS);

foreach ($html_files as $html_file) {
    $html_content = file_get_contents($html_file);

    $parts = explode(DIRECTORY_SEPARATOR, dirname($html_file));
    $page_path = end($parts);
    $output_file = $OUTPUT_PATH . '/' . $page_path . '.dat';

    $num_items_written = 0;
    $check_duplicates = [];

    // Use DOMDocument and DOMXPath for parsing
    libxml_use_internal_errors(true);
    $dom = new DOMDocument();
    $dom->loadHTML($html_content, LIBXML_NOERROR | LIBXML_NOWARNING);
    $xpath = new DOMXPath($dom);

    $class_selector = [];
    foreach ($CSS_CLASSES as $class) {
        $class_selector[] = "contains(concat(' ', normalize-space(@class), ' '), ' $class ')";
    }
    $selector = "//div[" . implode(' or ', $class_selector) . "]";

    $ofile = fopen($output_file, 'w');
    foreach ($xpath->query($selector) as $div) {
        foreach (['p', 'span', 'a'] as $tag) {
            foreach ($div->getElementsByTagName($tag) as $item) {
                $text = $item->textContent;
                if (!in_array($text, $check_duplicates) && !in_array($text, $SKIP_TEXT)) {
                    $check_duplicates[] = $text;
                    $text = preg_replace('/\s+/', ' ', $text);
                    $text = trim($text);
                    if ($text !== '') {
                        fwrite($ofile, $text . PHP_EOL);
                        $num_items_written += 1;
                    }
                }
            }
        }
    }
    fclose($ofile);

    if ($num_items_written == 0) {
        unlink($output_file);
    }
}
